File: windows/DisplayShot/Annotations/AnnotationHitTester.cs
using System.Windows;

namespace DisplayShot.Annotations;

/// <summary>Decides whether an annotation is touched by the eraser circle (centre p, radius r).</summary>
public static class AnnotationHitTester
{
    public static double DistanceToSegment(Point p, Point a, Point b)
    {
        var dx = b.X - a.X;
        var dy = b.Y - a.Y;
        var len2 = dx * dx + dy * dy;
        if (len2 <= 0) return (p - a).Length;
        var t = Math.Max(0, Math.Min(1, ((p.X - a.X) * dx + (p.Y - a.Y) * dy) / len2));
        return (p - new Point(a.X + t * dx, a.Y + t * dy)).Length;
    }

    private static bool Polyline(IReadOnlyList<Point> pts, Point p, double reach)
    {
        if (pts.Count == 0) return false;
        if (pts.Count == 1) return (pts[0] - p).Length <= reach;
        for (var i = 0; i < pts.Count - 1; i++)
            if (DistanceToSegment(p, pts[i], pts[i + 1]) <= reach) return true;
        return false;
    }

    private static Rect Inflated(Rect r, double d) { var c = r; c.Inflate(d, d); return c; }

    public static bool Hits(Annotation a, Point p, double r) => a switch
    {
        PenAnnotation pen => Polyline(pen.Points, p, r + pen.Stroke.Width / 2),
        MarkerAnnotation m => Polyline(m.Points, p, r + m.Stroke.Width / 2),
        LineAnnotation l => DistanceToSegment(p, l.From, l.To) <= r + Math.Max(l.Stroke.Width / 2, l.Stroke.Width * 1.6),
        ArrowAnnotation ar => DistanceToSegment(p, ar.From, ar.To) <= r + Math.Max(ar.Stroke.Width / 2, ar.Stroke.Width * 1.6),
        RectangleAnnotation rect => Outline(rect.Rect, p, r + rect.Stroke.Width / 2),
        TextAnnotation t => Inflated(new Rect(t.Origin, AnnotationRenderer.MeasureText(t)), r).Contains(p),
        EmojiAnnotation e => Inflated(e.Bounds, r).Contains(p),
        RedactAnnotation rd => Inflated(rd.Rect, r).Contains(p),
        _ => false,
    };

    private static bool Outline(Rect rect, Point p, double reach)
    {
        var tl = rect.TopLeft; var tr = rect.TopRight; var bl = rect.BottomLeft; var br = rect.BottomRight;
        return DistanceToSegment(p, tl, tr) <= reach || DistanceToSegment(p, tr, br) <= reach
            || DistanceToSegment(p, br, bl) <= reach || DistanceToSegment(p, bl, tl) <= reach;
    }
}

File: macos/Sources/DisplayShotKit/Annotations/AnnotationHitTester.swift
import AppKit

/// Decides whether an annotation is touched by the eraser circle (centre `p`, radius `r`).
enum AnnotationHitTester {
    static func hits(_ a: Annotation, circleAt p: CGPoint, radius r: CGFloat) -> Bool {
        switch a.kind {
        case .pen(let pts, let s), .marker(let pts, let s):
            let reach = r + s.width / 2
            guard pts.count > 1 else { return pts.first.map { $0.distance(to: p) <= reach } ?? false }
            for i in 0..<(pts.count - 1) where distance(from: p, toSegment: pts[i], pts[i + 1]) <= reach { return true }
            return false
        case .line(let a, let b, let s), .arrow(let a, let b, let s):
            return distance(from: p, toSegment: a, b) <= r + max(s.width / 2, s.width * 1.6)
        case .rectangle(let rect, let s):
            let reach = r + s.width / 2
            let tl = CGPoint(x: rect.minX, y: rect.minY), tr = CGPoint(x: rect.maxX, y: rect.minY)
            let bl = CGPoint(x: rect.minX, y: rect.maxY), br = CGPoint(x: rect.maxX, y: rect.maxY)
            return [(tl, tr), (tr, br), (br, bl), (bl, tl)].contains { distance(from: p, toSegment: $0.0, $0.1) <= reach }
        case .ellipse(let rect, let s):
            // Distance to the ellipse outline at the parametric angle of `p`.
            let a = rect.width / 2, b = rect.height / 2
            guard a > 0, b > 0 else { return false }
            let c = rect.center
            let angle = atan2((p.y - c.y) / b, (p.x - c.x) / a)
            let q = CGPoint(x: c.x + a * cos(angle), y: c.y + b * sin(angle))
            return p.distance(to: q) <= r + s.width / 2
        case .text(let t):
            let size = AnnotationRenderer.textSize(t)
            return CGRect(origin: t.origin, size: size).insetBy(dx: -r, dy: -r).contains(p)
        case .emoji(let e):
            return e.bounds.insetBy(dx: -r, dy: -r).contains(p)
        case .redact(let rect, _, _):
            return rect.insetBy(dx: -r, dy: -r).contains(p)
        }
    }
}
